spec/middleware_spec.rb

```diff
@@ -7,194 +7,238 @@
+      mw_auth_false = described_class.new(rack_app, should_authenticate_check: proc { false })
+      env = double
+      expect(mw_auth_false.mauth_client).not_to receive(:authentic?)
+      expect(rack_app).to receive(:call).with(env).and_return(res)
+      status, headers, body = mw_auth_false.call(env)
+      expect(200).to eq(status)
+      expect(['hello world']).to eq(body)
+    end
+
+    it 'authenticates if should_authenticate_check is omitted or indicates to' do
+      [nil, proc { |_env| true }].each do |should_authenticate_check|
+        mw_w_flag = described_class.new(rack_app, should_authenticate_check: should_authenticate_check)
+        env = { 'HTTP_X_MWS_AUTHENTICATION' => 'MWS foo:bar' }
+        expect(mw_w_flag.mauth_client).to receive(:authentic?).and_return(true)
+        expect(rack_app).to receive(:call).with(env.merge('mauth.app_uuid' => 'foo', 'mauth.authentic' => true)).and_return(res)
+        status, headers, body = mw_w_flag.call(env)
+        expect(status).to eq(200)
+        expect(body).to eq(['hello world'])
+      end
+    end
+
+    it 'returns 401 and does not call the app if authentication fails' do
+      expect(mw.mauth_client).to receive(:authentic?).and_return(false)
+      expect(rack_app).not_to receive(:call)
+      status, headers, body = mw.call({ 'REQUEST_METHOD' => 'GET' })
+      expect(401).to eq(status)
+      expect(body.join).to match(/Unauthorized/)
+    end
+
+    it 'returns 401 with no body if the request method is HEAD and authentication fails' do
+      expect(mw.mauth_client).to receive(:authentic?).and_return(false)
+      expect(rack_app).not_to receive(:call)
+      status, headers, body = mw.call({ 'REQUEST_METHOD' => 'HEAD' })
+      expect(headers["Content-Length"].to_i).to be > 0
+      expect(401).to eq(status)
+      expect([]).to eq(body)
+    end
+
+    it 'returns 500 and does not call the app if unable to authenticate' do
+      expect(mw.mauth_client).to receive(:authentic?).and_raise(MAuth::UnableToAuthenticateError.new(''))
+      expect(rack_app).not_to receive(:call)
+      status, headers, body = mw.call({ 'REQUEST_METHOD' => 'GET' })
+      expect(500).to eq(status)
+      expect(body.join).to match(/Could not determine request authenticity/)
+    end
+
+    context 'the AUTHENTICATE_WITH_ONLY_V2 flag is true and the request has only v1 headers' do
+      let(:authenticate_with_only_v2) { true }
+
+      it 'returns 401 with an informative message and does not call the app' do
+        env = { 'HTTP_X_MWS_AUTHENTICATION' => 'MWS foo:bar', 'REQUEST_METHOD' => 'GET' }
+        expect(mw.mauth_client).not_to receive(:authentic?)
+        expect(rack_app).not_to receive(:call)
+        status, headers, body = mw.call(env)
+        expect(401).to eq(status)
+        expect(headers['Content-Type']).to eq('application/json')
+        expect(JSON.parse(body.join)).to eq({
+            'type' => 'errors:mauth:missing_v2',
```

Comment:
Seems like too much indentation here since `({` and `})` are each on their own same line